runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -1961,6 +1961,376 @@ J9::Z::TreeEvaluator::inlineUTF16BEEncode(TR::Node *node, TR::CodeGenerator *cg)
+   TR::LabelSymbol* foldBy1Loop = generateLabelSymbol(cg);
+   TR::LabelSymbol* foldBy1LoopEntry = generateLabelSymbol(cg);
+   TR::LabelSymbol* finalReduction = generateLabelSymbol(cg);
+   TR::LabelSymbol* callJava = generateLabelSymbol(cg);
+   TR::LabelSymbol* endICF = generateLabelSymbol(cg);
+
+   // For debug annototations
+   TR::Instruction *cursor;
+   TR_Debug *debugObj = cg->getDebug();
+
+   // If less than 16B of input, branch to bytewise path
+   // The vectorized path only works for multiples of 16B
+   generateRIInstruction(cg, TR::InstOpCode::CGHI, node, len, 16);
+   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BL, node, callJava);
+   if (debugObj)
+      debugObj->addInstructionComment(cursor, "Jump to OOL call to original Java implementation if < 16B");
+
+   // The CRC-32C vector constant block contains reduction constants to fold and process particular
+   // chunks of the input data stream in parallel.
+   //
+   // The constants are precomputed according to these definitions:
+   //
+   //	R1 = [(x^(4*128+32) mod P'(x) << 32)]' << 1
+   //	R2 = [(x^(4*128-32) mod P'(x) << 32)]' << 1
+   //	R3 = [(x^(128+32) mod P'(x) << 32)]'   << 1
+   //	R4 = [(x^(128-32) mod P'(x) << 32)]'   << 1
+   //	R5 = [(x^64 mod P'(x) << 32)]'	      << 1
+   //	R6 = [(x^32 mod P'(x) << 32)]'	      << 1
+   //
+   //	The bitreflected Barret reduction constant, u', is defined as the bit reversal of floor(x^64 / P(x))
+   //	where P(x) is the polynomial in the normal domain and the P'(x) is the polynomial in the reversed (bitreflected) domain.
+   //
+   // CRC-32C (Castagnoli) polynomials:
+   //
+   //	P(x)  = 0x1EDC6F41
+   //	P'(x) = 0x82F63B78
+
+   // Array of 16 rather than 12 because data snippets on Z must be a power of 2 size - See OMR issue #1815
+   uint64_t crc32cVectorConstants[16] =
+      {
+      0x0F0E0D0C0B0A0908ull, 0x0706050403020100ull,   // BE->LE
+      0x000000009E4ADDF8ull, 0x00000000740EEF02ull,   // R2, R1
+      0x000000014CD00Bd6ull, 0x00000000F20C0dFEull,   // R4, R3
+      0x0000000000000000ull, 0x00000000DD45AAB8ull,   // R5
+      0x0000000000000000ull, 0x00000000DEA713F1ull,   // u'
+      0x0000000000000000ull, 0x0000000105ec76f0ull,   // P'(x) << 1
+      };
+
+   TR::Snippet* constants = cg->findOrCreateConstant(node, crc32cVectorConstants, 128);
+   generateRILInstruction(cg, TR::InstOpCode::LARL, node, tablePtr, constants);
+   cursor = generateVRSaInstruction(cg, TR::InstOpCode::VLM, node, vConstPermLE2BE, vConstCRCPoly, generateS390MemoryReference(tablePtr, 0, cg), 0);
+   if (debugObj)
+      debugObj->addInstructionComment(cursor, "Populate vectors with CRC-32C reduction constants");
+
+   // Load the initial CRC value into the rightmost word of vScratch
+   generateVRIaInstruction(cg, TR::InstOpCode::VGBM, node, vScratch, 0, 0);
+   generateVRSbInstruction(cg, TR::InstOpCode::VLVG, node, vScratch, crc, generateS390MemoryReference(3, cg), 2);
```

Comment:
This is ultimately initializing the `crc` with 0 right? If that is the case, can we use `XGR`?

Reply:
This is initializing `vScratch` with 0, then loading the `crc` into `vScratch`